src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/Compression/WebSocketDeflater.cs

```diff
@@ -0,0 +1,229 @@
+        {
+            Debug.Assert(_stream is not null);
+
+            fixed (byte* fixedInput = input)
+            fixed (byte* fixedOutput = output)
+            {
+                _stream.NextIn = (IntPtr)fixedInput;
+                _stream.AvailIn = (uint)input.Length;
+
+                _stream.NextOut = (IntPtr)fixedOutput;
+                _stream.AvailOut = (uint)output.Length;
+
+                // If flush is set to Z_BLOCK, a deflate block is completed
+                // and emitted, as for Z_SYNC_FLUSH, but the output
+                // is not aligned on a byte boundary, and up to seven bits
+                // of the current block are held to be written as the next byte after
+                // the next deflate block is completed.
+                var errorCode = Deflate(_stream, (FlushCode)5/*Z_BLOCK*/);
+
+                consumed = input.Length - (int)_stream.AvailIn;
+                written = output.Length - (int)_stream.AvailOut;
+
+                needsMoreBuffer = errorCode == ErrorCode.BufError;
+            }
+        }
+
+        private unsafe int UnsafeFlush(Span<byte> output)
+        {
+            Debug.Assert(_stream is not null);
+            Debug.Assert(_stream.AvailIn == 0);
+            Debug.Assert(output.Length >= 6, "We neede at least 6 bytes guarantee the completion of the deflate block.");
+
+            fixed (byte* fixedOutput = output)
+            {
+                _stream.NextIn = IntPtr.Zero;
+                _stream.AvailIn = 0;
+
+                _stream.NextOut = (IntPtr)fixedOutput;
+                _stream.AvailOut = (uint)output.Length;
+
+                ErrorCode errorCode = Deflate(_stream, (FlushCode)3/*Z_FULL_FLUSH*/);
+                int writtenBytes = output.Length - (int)_stream.AvailOut;
+                Debug.Assert(errorCode == ErrorCode.Ok);
+
+                return writtenBytes;
+            }
+        }
+
+        private static ErrorCode Deflate(ZLibStreamHandle stream, FlushCode flushCode)
+        {
+            ErrorCode errorCode;
+            try
+            {
+                errorCode = stream.Deflate(flushCode);
+            }
+            catch (Exception cause)
+            {
+                throw new WebSocketException(SR.ZLibErrorDLLLoadError, cause);
+            }
+
+            switch (errorCode)
+            {
+                case ErrorCode.Ok:
+                case ErrorCode.StreamEnd:
+                    return errorCode;
+
+                case ErrorCode.BufError:
+                    return errorCode;  // This is a recoverable error
+
+                case ErrorCode.StreamError:
+                    throw new WebSocketException(SR.ZLibErrorInconsistentStream);
+
+                default:
+                    throw new WebSocketException(string.Format(SR.ZLibErrorUnexpected, (int)errorCode));
+            }
```

Comment:
Can reduce code gen here:

```suggestion
           
            if (errorCode is ErrorCode.Ok or ErrorCode.StreamEnd or ErrorCode.BufError)
            {
                // BufError is a recoverable error
                return errorCode;
            }
            else
            {
                    string message = errorCode == ErrorCode.StreamError
                        ? SR.ZLibErrorInconsistentStream
                        : string.Format(SR.ZLibErrorUnexpected, (int)errorCode);

                    throw new WebSocketException(message);
            }
```